### ISSUE TITLE
How do we handle transactional file operations?

### ISSUE THREAD

**Issue by dstufft** [pypa/warehouse#473]
Warehouse (via [pyramid_tm](https://pypi.python.org/pypi/pyramid_tm/) and [transaction[(https://pypi.python.org/pypi/transaction/)) has a generic transaction system that any data store can implement and the entire set of data storage engines will be commited or rollback'd at once. Currently this is only in use for the database access by plugging in the transaction system that comes for free from a SQL database into it.

However this doesn't solve the problem of how we manage operations to the file system, particularly state modifying operations like write and delete. This ticket is here to brainstorm ideas for how to implement file storage in Warehouse in such a way that it is tied to the generic transaction system.

The transaction system has available the following hooks for implementing the transaction protocol:

Hook | Description
---- | -----------
``abort`` | Outside of the transaction protocol just abort the current transaction, throw away all uncommited changes and reset back to zero.
``tpc_begin`` | Called all the start of the transaction, any work that needs to be done to prepare a system for actually getting ready to do the transaction (such as flushing any pending changes).
``commit`` | Prepare the transaction, after this is called changes should be ready to be made permanent but should not yet be permanent.
``tpc_vote`` | Last chance to ensure that a transaction can be committed before all of the available data storage engines are told to commit.
``tpc_finish`` | Actually commit the transaction to permanent storage, at this point all data storage engines voted yes and it's time to save the data.
``tpc_abort`` | Called instead of ``tpc_finish`` if any data storage engine voted no, should throw away all pending changes and end the transaction.

In addition, callbacks can also be registered for before and after commits if we want to delay an action until right before or after committing actually happens without actually participating in the transaction itself.

Currently in PyPI legacy we more or less ignore the fact that files are not part of the transaction which means that if a transaction fails to commit our file system will be in a state that is inconsistent with the state of the database. This hasn't been a *major* problem to my knowledge because we're very light on writes but very heavy on reads and transactions generally don't fail often. However I think this is going to be a bigger problem with Warehouse because we've solved some concurrency problems by switching our transaction isolation from ``READ COMMITTED`` to ``SERIALIZABLE`` (which we handle serialization failures by simply implicitly re-executing the view function).

Basic Constraints:

* Almost all writes are write once per filename, however it should be able to generally do the right thing in the case of an update. Updates don't need to be perfect though because an update requires manual intervention from an administrator to allow anyways.
* Deletes are rare but should be sanely handled and are generally "forever" without admin intervention. It would be acceptable I think for deletes to be ignored completely as long as updates can write over them.
* All files have a unique row in the SQL database which is unique on the filename itself and this row is trusted as the source of "truth".

#### Idea 1: PyPI Legacy Style

The crux of this idea is that not worrying about it has served PyPI legacy fine for a long time. So we'd simply write data in the middle of the view and if the transaction rolls back then the object store will just have extra data in it. This would require that the view itself can overwrite data in the object store and would more or less ignore whatever is there and just blindly overwrite it. The upside is this is very simple to implement, but the downside is that if people navigate directly to a file URL they may get a file that doesn't exist anywhere else in the system and we might end up overwriting the file multiple times if we need to retry the view. To be clear, this would write the file data *before* committing the transaction.

#### Idea 2: PyPI Legacy Style but After

This is basically the same thing as Idea 1, except delaying the write until after the transaction commits. This would reduce the amount of writes we need in the case of a retry however it would exchange the inconsistency, instead of having changes that exist in the file storage but not in database, we'd have changes that exist in the database but not in file storage if something prevented it from saving.

#### Idea 3: PostgreSQL Advisory Locks

This would more or less be implementing the commit protocol but using the PostgreSQL advisory locks. We'd need some way to turn a filename into a 64 bit integer but this would then work with a transaction system that is well understood already. 

#### Idea 4: Redis as a Lock

This would basically be the same thing as Idea 3, except we'd use redis as a distributed lock. Redis makes it easy to set a key only if it doesn't exist but it doesn't make it easy to release a lock if something crashes hard. We could handle this by having a timeout on the lock itself though.

#### Idea 5: Consul as a Lock

Basically idea 3 and 4 again but this time using consul (or etcd, or ZK, or whatever) as a lock. These systems are actually designed to be a lock but running them is a lot more complicated (generally they require 3+ nodes) although we are already running consul in the PSF infrastructure but I'd greatly prefer not to have applications themselves call into that. We could run a single node consul instance to handle locking for this but that feels pretty heavy weight to have a whole service dedicated to locking when we already have PostgreSQL and Redis.

Does anyone have any other thoughts? Currently I'm bouncing between 3, 4, and 1. I don't like 2 because extra data in the file storage API doesn't feel as bad to me as having data inside of the database that doesn't exist in file storage. 5 I don't like because if feels way too heavy weight for this use case.

**Comment by vladimir-v-diaz** [comment]
"Almost all writes are write once per filename, however it should be able to generally do the right thing in the case of an update. Updates don't need to be perfect though because an update requires manual intervention from an administrator to allow anyways."

I'm not exactly sure what you mean by an update to a file.

Do administrators manually update hashes listed elsewhere (and possibly re-sign packages that they control) after modifying a file?  What about updates to a file that has been signed by a developer?

**Comment by dstufft** [comment]
So currently PyPI only allows you to upload a particular filename once, any attempts to delete that file and then re-upload it are denied. However in certain, special cases, a PyPI administrator can release the block on a re-upload for a particular filename and in that case an author can upload a new file in place of a previous file.

In other words, the common case is that there will only ever be 1 unique set of bytes per filename but that in certain rare cases there *might* be a situation where the author is able to replace that.

**Comment by domenkozar** [comment]
http://acidfs.readthedocs.org/en/latest/

**Comment by dstufft** [comment]
This is no longer a problem, filenames are written to S3 using a path that includes a hash of the contents of the file, so changing the file contents by definition changes the location on "disk". This means we can easily write the file to S3 prior to committing the transaction and if the transaction fails, it's no big deal because it just means we have an extraneous file laying around. Similarly deletes can just be ignored (or implemented using a sort of offline garbage collector).

So basically, we worked around the problem.
